// redux-thunk-data/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setData = exports.selectEntityByKeyAndId = exports.selectEntitiesByKeyAndJoin = exports.resetData = exports.reinitializeData = exports.mergeData = exports.getSuccessState = exports.getStateKeyFromUrl = exports.getStateKeyFromConfig = exports.getStateKeyFromApiPath = exports.getNormalizedMergedState = exports.getNormalizedDeletedState = exports.fetchData = exports.deleteData = exports.createDataReducer = exports.assignData = exports.SET_DATA = exports.RESET_DATA = exports.REINITIALIZE_DATA = exports.MERGE_DATA = exports.DELETE_DATA = exports.ASSIGN_DATA = exports.SERVER_ERROR = exports.API_ERROR = undefined;

var _requestData = require('./requestData');

Object.keys(_requestData).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _requestData[key];
    }
  });
});

var _fetchNormalizeData = require('fetch-normalize-data');

exports.API_ERROR = _fetchNormalizeData.API_ERROR;
exports.SERVER_ERROR = _fetchNormalizeData.SERVER_ERROR;
exports.ASSIGN_DATA = _fetchNormalizeData.ASSIGN_DATA;
exports.DELETE_DATA = _fetchNormalizeData.DELETE_DATA;
exports.MERGE_DATA = _fetchNormalizeData.MERGE_DATA;
exports.REINITIALIZE_DATA = _fetchNormalizeData.REINITIALIZE_DATA;
exports.RESET_DATA = _fetchNormalizeData.RESET_DATA;
exports.SET_DATA = _fetchNormalizeData.SET_DATA;
exports.assignData = _fetchNormalizeData.assignData;
exports.createDataReducer = _fetchNormalizeData.createDataReducer;
exports.deleteData = _fetchNormalizeData.deleteData;
exports.fetchData = _fetchNormalizeData.fetchData;
exports.getNormalizedDeletedState = _fetchNormalizeData.getNormalizedDeletedState;
exports.getNormalizedMergedState = _fetchNormalizeData.getNormalizedMergedState;
exports.getStateKeyFromApiPath = _fetchNormalizeData.getStateKeyFromApiPath;
exports.getStateKeyFromConfig = _fetchNormalizeData.getStateKeyFromConfig;
exports.getStateKeyFromUrl = _fetchNormalizeData.getStateKeyFromUrl;
exports.getSuccessState = _fetchNormalizeData.getSuccessState;
exports.mergeData = _fetchNormalizeData.mergeData;
exports.reinitializeData = _fetchNormalizeData.reinitializeData;
exports.resetData = _fetchNormalizeData.resetData;
exports.selectEntitiesByKeyAndJoin = _fetchNormalizeData.selectEntitiesByKeyAndJoin;
exports.selectEntityByKeyAndId = _fetchNormalizeData.selectEntityByKeyAndId;
exports.setData = _fetchNormalizeData.setData;
